src/scheduler/scheduler.ts

```diff
@@ -0,0 +1,183 @@
+            if (ro.schedulerConfig) {
+                const loop = new DetachableLoop(this.dbosExec, ro.schedulerConfig.crontab ?? '* * * * *', ro);
+                this.schedLoops.push(loop);
+                this.schedTasks.push(loop.startLoop());
+            }
+        }
+    }
+  
+    async destroyScheduler() {
+        for (const l of this.schedLoops) {
+            await l.stopLoop();
+        }
+        this.schedLoops = [];
+        await Promise.all(this.schedTasks);
+        this.schedTasks = [];
+    }
+  
+    logRegisteredSchedulerEndpoints() {
+        const logger = this.dbosExec.logger;
+        logger.info("Scheduled endpoints:");
+        this.dbosExec.registeredOperations.forEach((registeredOperation) => {
+            const ro = registeredOperation as SchedulerRegistration<unknown, unknown[], unknown>;
+            if (ro.schedulerConfig) {
+                logger.info(`    ${ro.name} @ ${ro.schedulerConfig.crontab ?? '* * * * *'}; ${ro.schedulerConfig.mode ?? 'Exactly Once Per Interval'}`);
+            }
+        });
+    }
+}
+
+class DetachableLoop {
+    private isRunning: boolean = false;
+    private resolveCompletion?: (value: void | PromiseLike<void>) => void;
+    private interruptResolve?: () => void;
+    private lastExec: Date;
+    private timeMatcher: TimeMatcher;
+
+    constructor(readonly dbosExec: DBOSExecutor, readonly crontab: string, readonly mtd: SchedulerRegistration<unknown, unknown[], unknown>) {
+        this.lastExec = new Date();
+        this.lastExec.setMilliseconds(0);
+        this.timeMatcher = new TimeMatcher(crontab);
+    }
+
+    async startLoop(): Promise<void> {
+        // See if the exec time is available in durable storage...
+        const lasttm = await this.dbosExec.systemDatabase.getLastScheduledTime(this.mtd.name);
+        if (lasttm) {
+            this.lastExec = new Date(lasttm);
+        }
+
+        this.isRunning = true;
+        while (this.isRunning) {
+            const nextExecTime = this.timeMatcher.nextWakeupTime(this.lastExec);
+            const sleepTime = nextExecTime.getTime() - new Date().getTime();
+            console.log(`Loop iteration with sleep time: ${sleepTime}ms`);
+
+            if (sleepTime > 0) {
+                // Wait for either the timeout or an interruption
+                await Promise.race([
+                    this.sleep(sleepTime),
+                    new Promise<void>((_, reject) => this.interruptResolve = reject)
+                ])
+                .catch(); // Interrupt sleep throws
+            }
+
+            if (!this.isRunning) {
+                break;
+            }
+
+            // Check crontab
+            if (!this.timeMatcher.match(nextExecTime)) {
+                this.lastExec = nextExecTime;
+                continue;
+            }
+
+            // Init workflow
+            const workflowUUID = `sched-${this.mtd.name}-${nextExecTime.toISOString()}`;
```

Comment:
This seems vulnerable to clock skew. For the same scheduled workflow, 2 nodes could generate a different `nextWakeupTime` (the value under `nextExecTime`) for the same interval. So we would have 2 different workflow UUID for the same workflow.

Arguably, this is only a problem for sub seconds intervals if the nodes are properly synchronized. (Cases where a node's clock is broken can happen, but quite less likely.)

Reply:
No, the exec time is the scheduled time, they'll both calculate the same time that it is _supposed_ to run.  It is a race between the nodes... if you had 1000 nodes and a 1-sec interval this design is terrible because they all pound the system DB every second seeing if anyone did the work.

Reply:
I understand: even tho all local nodes are seeded from the local clock (the first `lastExec` is using `Date()`), they are all offset to the expected schedule time. Interestingly this `nextWakeupTime` seems to always return at most +1h?

So multiple nodes, even with skew, should always see the same scheduled time.